File: python/activities/checkout.py
"""External checkout steps used by ``CheckoutWorkflow``.

These small Activities stand in for airline, hotel, and attraction APIs. Their
reservation IDs are idempotent so an Activity retry cannot create a duplicate.
The hotel failure is enabled by default for the live compensation demo; set
``CHECKOUT_FAIL_HOTEL=false`` to exercise the successful checkout path.
"""

import hashlib
import time

from temporalio import activity
from temporalio.exceptions import ApplicationError

import config
from models.types import (
    CheckoutRequest,
    CheckoutReservation,
    CheckoutStepRequest,
)

from . import db


def _pause() -> None:
    """Make the individual steps easy to see on the Temporal timeline."""
    if config.CHECKOUT_STEP_DELAY_SECONDS > 0:
        time.sleep(config.CHECKOUT_STEP_DELAY_SECONDS)


def _reservation(req: CheckoutStepRequest) -> CheckoutReservation:
    # A real integration would pass this key to the provider. Deriving it from
    # stable business identifiers gives this demo the same retry-safe behavior.
    raw = f"{req.account_key}:{req.item.kind}:{req.item.ref_id}".encode()
    suffix = hashlib.sha256(raw).hexdigest()[:10].upper()
    prefix = {"flight": "FLT", "hotel": "HTL", "activity": "ACT"}[req.item.kind]
    return CheckoutReservation(
        kind=req.item.kind,
        ref_id=req.item.ref_id,
        title=req.item.title,
        reservation_id=f"{prefix}-{suffix}",
    )


@activity.defn
def book_flight(req: CheckoutStepRequest) -> CheckoutReservation:
    _pause()
    return _reservation(req)


@activity.defn
def book_hotel(req: CheckoutStepRequest) -> CheckoutReservation:
    _pause()
    if config.CHECKOUT_FAIL_HOTEL:
        raise ApplicationError(
            "Hotel booking failed — the supplier returned no availability "
            "(injected demo failure).",
            type="HotelBookingFailed",
            non_retryable=True,
        )
    return _reservation(req)


@activity.defn
def book_activity(req: CheckoutStepRequest) -> CheckoutReservation:
    _pause()
    return _reservation(req)


def _cancel(reservation: CheckoutReservation) -> CheckoutReservation:
    _pause()
    return reservation.model_copy(update={"status": "cancelled"})


@activity.defn
def cancel_flight(reservation: CheckoutReservation) -> CheckoutReservation:
    return _cancel(reservation)


@activity.defn
def cancel_hotel(reservation: CheckoutReservation) -> CheckoutReservation:
    return _cancel(reservation)


@activity.defn
def cancel_activity(reservation: CheckoutReservation) -> CheckoutReservation:
    return _cancel(reservation)


@activity.defn
def finalize_checkout(req: CheckoutRequest) -> dict:
    """Persist the completed trip only after every reservation succeeds."""
    items = [item.model_dump() for item in req.items]
    already = db.items_already_booked(req.account_key, items)
    if already:
        names = ", ".join(f'"{item["title"]}"' for item in already)
        raise ApplicationError(
            f"Booking declined — this trip already includes {names}.",
            type="BookingDeclined",
            non_retryable=True,
        )
    return db.record_booking(req.account_key, items, req.summary)

File: typescript/src/activities/checkout.ts
// External checkout steps used by CheckoutWorkflow. These are intentionally
// tiny stand-ins for airline, hotel, and attraction APIs. Stable reservation
// IDs make retries idempotent, just as a provider idempotency key would.

import { createHash } from 'crypto';
import { ApplicationFailure } from '@temporalio/common';

import * as config from '../config';
import type {
  CheckoutRequest,
  CheckoutReservation,
  CheckoutStepRequest,
} from '../types';
import * as db from './db';

async function pause(): Promise<void> {
  if (config.CHECKOUT_STEP_DELAY_MS > 0) {
    await new Promise((resolve) => setTimeout(resolve, config.CHECKOUT_STEP_DELAY_MS));
  }
}

function reservation(req: CheckoutStepRequest): CheckoutReservation {
  const raw = `${req.account_key}:${req.item.kind}:${req.item.ref_id}`;
  const suffix = createHash('sha256').update(raw).digest('hex').slice(0, 10).toUpperCase();
  const prefix = { flight: 'FLT', hotel: 'HTL', activity: 'ACT' }[req.item.kind];
  return {
    kind: req.item.kind,
    ref_id: req.item.ref_id,
    title: req.item.title,
    reservation_id: `${prefix}-${suffix}`,
    status: 'booked',
  };
}

export async function book_flight(req: CheckoutStepRequest): Promise<CheckoutReservation> {
  await pause();
  return reservation(req);
}

export async function book_hotel(req: CheckoutStepRequest): Promise<CheckoutReservation> {
  await pause();
  if (config.CHECKOUT_FAIL_HOTEL) {
    throw ApplicationFailure.create({
      message:
        'Hotel booking failed — the supplier returned no availability (injected demo failure).',
      type: 'HotelBookingFailed',
      nonRetryable: true,
    });
  }
  return reservation(req);
}

export async function book_activity(req: CheckoutStepRequest): Promise<CheckoutReservation> {
  await pause();
  return reservation(req);
}

async function cancel(res: CheckoutReservation): Promise<CheckoutReservation> {
  await pause();
  return { ...res, status: 'cancelled' };
}

export async function cancel_flight(res: CheckoutReservation): Promise<CheckoutReservation> {
  return cancel(res);
}

export async function cancel_hotel(res: CheckoutReservation): Promise<CheckoutReservation> {
  return cancel(res);
}

export async function cancel_activity(res: CheckoutReservation): Promise<CheckoutReservation> {
  return cancel(res);
}

export async function finalize_checkout(req: CheckoutRequest): Promise<Record<string, unknown>> {
  const already = await db.itemsAlreadyBooked(req.account_key, req.items);
  if (already.length) {
    const names = already.map((item) => `"${item.title}"`).join(', ');
    throw ApplicationFailure.create({
      message: `Booking declined — this trip already includes ${names}.`,
      type: 'BookingDeclined',
      nonRetryable: true,
    });
  }
  return db.recordBooking(req.account_key, req.items, req.summary);
}
